engine/consensus/sealing/engine.go

```diff
@@ -279,37 +318,59 @@ func (e *Engine) ProcessLocal(event interface{}) error {
 	return e.Process(e.me.NodeID(), event)
 }
 
-// HandleReceipt pipes explicitly requested receipts to the process function.
-// Receipts can come from this function or the receipt provider setup in the
-// engine constructor.
-func (e *Engine) HandleReceipt(originID flow.Identifier, receipt flow.Entity) {
-	e.log.Debug().Msg("received receipt from requester engine")
-
-	err := e.Process(originID, receipt)
-	if err != nil {
-		e.log.Error().Err(err).Hex("origin", originID[:]).Msg("could not process receipt")
-	}
-}
-
 // Ready returns a ready channel that is closed once the engine has fully
 // started. For the propagation engine, we consider the engine up and running
 // upon initialization.
 func (e *Engine) Ready() <-chan struct{} {
-	var wg sync.WaitGroup
-	wg.Add(2)
-	e.unit.Launch(func() {
-		wg.Done()
-		e.processEvents()
-	})
-	e.unit.Launch(func() {
-		wg.Done()
-		e.consumeEvents()
-	})
-	return e.unit.Ready(func() {
-		wg.Wait()
-	})
+	// launch as many workers as we need
+	for i := 0; i < defaultSealingEngineWorkers; i++ {
+		e.unit.Launch(e.loop)
+	}
+	return e.unit.Ready()
 }
 
 func (e *Engine) Done() <-chan struct{} {
 	return e.unit.Done()
 }
+
+// OnFinalizedBlock implements the `OnFinalizedBlock` callback from the `hotstuff.FinalizationConsumer`
+//  (1) Informs sealing.Core about finalization of respective block.
+// CAUTION: the input to this callback is treated as trusted; precautions should be taken that messages
+// from external nodes cannot be considered as inputs to this function
+func (e *Engine) OnFinalizedBlock(finalizedBlockID flow.Identifier) {
+	e.pendingFinalizationEvents.Push(finalizedBlockID)
```

Comment:
This approach works, just provide some idea for optimization:

The finalized block ID is used to prune the collection tree by last finalized height. Here we are pushing each finalized block into a fifo queue, and the consumer of the fifoqueue (the sealing core) will need to process each message from the queue. I don't think it's necessary.

In fact, we only need to process the _latest_ message, instead of every message. 

For example, if we are finalizing from block 101 to 120. We will push 20 events to the fifoqueue, and the sealing core will prune the collection tree 20 times from height from 101 to 120. Instead, we only need to prune once, just prune by 120. This happens when the sealing core was busy, and the fifoqueue has buffered 20 messages.

Another case is if we the fifoqueue reaches the capacity, we will drop messages, and if finalization halts, then we won't be able to process the latest finalized block, because it has been dropped. But I guess it's less of an issue, as we assume the finalization won't halt.